airflow/traces/tracer.py

```diff
@@ -0,0 +1,256 @@
+
+    return otel_gen_context(trace_id, span_id)
+
+
+def gen_links_from_kv_list(list):
+    """Generate links from kv list of {trace_id:int, span_id:int}."""
+    from airflow.traces.otel_tracer import gen_links_from_kv_list
+
+    return gen_links_from_kv_list(list)
+
+
+def span(func):
+    """Decorate a function with span."""
+
+    def wrapper(*args, **kwargs):
+        func_name = func.__name__
+        qual_name = func.__qualname__
+        module_name = func.__module__
+        if "." in qual_name:
+            component = f"{qual_name.rsplit('.', 1)[0]}"
+        else:
+            component = module_name
+        with Trace.start_span(span_name=func_name, component=component):
+            return func(*args, **kwargs)
+
+    return wrapper
+
+
+class DummyContext:
+    """If no Tracer is configured, DummyContext is used as a fallback."""
+
+    def __init__(self):
+        self.trace_id = 1
+
+
+class DummySpan:
+    """If no Tracer is configured, DummySpan is used as a fallback."""
+
+    def __enter__(self):
+        """Enter."""
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        """Exit."""
+        pass
+
+    def __call__(self, obj):
+        """Call."""
+        return obj
+
+    def get_span_context(self):
+        """Get span context."""
+        return DUMMY_CTX
+
+    def set_attribute(self, key, value) -> None:
+        """Set an attribute to the span."""
+        pass
+
+    def set_attributes(self, attributes) -> None:
+        """Set multiple attributes at once."""
+        pass
+
+    def add_event(
+        self,
+        name: str,
+        attributes=None,
+        timestamp: int | None = None,
+    ) -> None:
+        """Add event to span."""
+        pass
+
+    def add_link(
+        self,
+        context: typing.Any,
+        attributes=None,
+    ) -> None:
+        """Add link to the span."""
+        pass
+
+    def end(self, end_time=None, *args, **kwargs) -> None:
+        """End."""
+        pass
+
+
+DUMMY_SPAN = DummySpan()
+DUMMY_CTX = DummyContext()
+
+
+class Tracer(Protocol):
+    """This class is only used for TypeChecking (for IDEs, mypy, etc)."""
+
+    instance: Tracer | DummyTrace | None = None
+
+    @classmethod
+    def get_tracer(cls, component):
+        """Get a tracer."""
+        raise NotImplementedError()
+
+    @classmethod
+    def start_span(
+        cls,
+        span_name: str,
+        component: str | None = None,
+        parent_sc=None,
+        span_id=None,
+        links=None,
+        start_time=None,
+    ):
+        """Start a span."""
+        raise NotImplementedError()
+
+    @classmethod
+    def use_span(cls, span):
+        """Use a span as current."""
+        raise NotImplementedError()
+
+    @classmethod
+    def get_current_span(self):
+        raise NotImplementedError()
+
+    @classmethod
+    def start_span_from_dagrun(cls, dagrun, span_name=None, service_name=None, component=None, links=None):
```

Comment:
Here and elsewhere:   Just my personal preference, feel free to resolve and ignore.  If you add a comma after the last arg, the static checks will force it to one per line like `start_span()` has them above.